### DISCUSSION TITLE
What makes a package !uwp?

### DISCUSSION THREAD

**Discussion by KRJackLee** [microsoft/vcpkg#39472] [Q&A] [answered]
Greetings. Recently I have been working on something that data read & parse is a performance bottleneck, like csv. High performance is the goal of this project. I haven't got time to prepare a sperate Linux environment so I've been using msvc+cmake for now. With vcpkg to manage packages things are working fine so far, until I am trying packages like `dataframe` and `vincentlaucsb-csv-parser`, who claims to have prevailing performance in their github repositories. However, both of them are !uwp. I haven't realized it until `vcpkg install dataframe:x64-windows`, only to find that `dataframe` seems to use cxx23 which vs2022 doesn't have full support of? At first, I thought that's the reason `dataframe` is !uwp, as it's in cxx23 now. However, `vincentlaucsb-csv-parser` claims to use cxx20, yet !uwp also. `vcpkg install vincentlaucsb-csv-parser:x64-windows` doesn't seem to find a valid installation, too. 
So, my questions:
1. What makes these packages !uwp? cxx standards, components they are using, syntax that msvc or visual studio can't comprehend?
2. Can anyone recommend packages for fast data read & parse to memory, ideally, parsing to a custom data struct? Or, recommendation  in writing a wheel to accomplish it? 
I am currently working on a 400MB file, around 1,050,000 x 9. My previous custom wheel took 360+s in read & parse. I've also tried `rapidcsv` before. Reading took 80+s and parsing took 140+s, while `pandas.read_csv` took 4.8+s in reading. I don't expect read optimization to the degree of `pandas`, but the whole program afterwards took way, way smaller runtime than data read & parse, thus making it a bottleneck.

**Comment by dg0yt** [comment]
> 1. What makes these packages !uwp?

The typical reason is the use of Windows APIs which are not available for UWP. That platform is much more restrictive. For some packages, these restrictions only prevent the build of tools.

You may try to install with `--allow-unsupported`. Either it wail fail, so that you can see the error, ot the original problem was fixed.

**Comment by KRJackLee** [comment]
btw, is there any explanation for all those triplets? I mean, for example, is uwp equal to all windows triplets, including x86-
windows, x64-windows, etc.?

**Reply by dg0yt** [reply, marked as answer]
They are not equal. You can check the settings in each triplet file in `triplets/`.  
https://learn.microsoft.com/vcpkg/get_started/overview#ports-and-triplets  
https://learn.microsoft.com/vcpkg/concepts/triplets

**Comment by KRJackLee** [comment]
Okay, after some digging, I realized what uwp is, the magnet sticker in the Start Menu since win 8, in short. Then it's almost clear why most packages don't support uwp, XD. Then I guess dataframe incompatibility comes from msvc not fully embracing cxx 23 then.
